From 48de6e2d2bc16e3b1640cd66ae47aab516a66cc5 Mon Sep 17 00:00:00 2001
From: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
Date: Thu, 24 Sep 2015 03:00:24 +0200
Subject: [PATCH] EFI/early: implement /noexit to not ExitBootServices
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Organization: Invisible Things Lab
Cc: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

The /noexit will inhibit Xen from calling ExitBootServices. This allows
on Lenovo ThinkPad x230 to use GetNextVariableName in 1-1 mapping mode.

Signed-off-by: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
Backported-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
Original patch:
http://lists.xen.org/archives/html/xen-devel/2015-04/msg02012.html

Signed-off-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
---
 xen/arch/x86/efi/boot.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/xen/arch/x86/efi/boot.c b/xen/arch/x86/efi/boot.c
index 011e443..6dd91d8 100644
--- a/xen/arch/x86/efi/boot.c
+++ b/xen/arch/x86/efi/boot.c
@@ -795,6 +795,7 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SYSTEM_TABLE *SystemTable)
     union string section = { NULL }, name;
     u64 efer;
     bool_t base_video = 0, retry;
+    bool_t exit_boot_services = 1;
 
     efi_ih = ImageHandle;
     efi_bs = SystemTable->BootServices;
@@ -847,6 +848,8 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SYSTEM_TABLE *SystemTable)
                 base_video = 1;
             else if ( wstrcmp(ptr + 1, L"mapbs") == 0 )
                 map_bs = 1;
+            else if ( wstrcmp(ptr + 1, L"noexit") == 0 )
+                exit_boot_services = 0;
             else if ( wstrncmp(ptr + 1, L"cfg=", 4) == 0 )
                 cfg_file_name = ptr + 5;
             else if ( i + 1 < argc && wstrcmp(ptr + 1, L"cfg") == 0 )
@@ -857,6 +860,7 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SYSTEM_TABLE *SystemTable)
                 PrintStr(L"Xen EFI Loader options:\r\n");
                 PrintStr(L"-basevideo   retain current video mode\r\n");
                 PrintStr(L"-mapbs       map EfiBootServices{Code,Data}\r\n");
+                PrintStr(L"-noexit      Don't call ExitBootServices\r\n");
                 PrintStr(L"-cfg=<file>  specify configuration file\r\n");
                 PrintStr(L"-help, -?    display this help\r\n");
                 blexit(NULL);
@@ -1478,8 +1482,11 @@ efi_start(EFI_HANDLE ImageHandle, EFI_SYSTEM_TABLE *SystemTable)
             relocate_trampoline(cfg.addr);
         }
 
-        status = SystemTable->BootServices->ExitBootServices(ImageHandle,
-                                                             map_key);
+        if (exit_boot_services)
+            status = SystemTable->BootServices->ExitBootServices(ImageHandle,
+                                                                 map_key);
+        else
+            status = 0;
         efi_bs = NULL;
         if ( status != EFI_INVALID_PARAMETER || retry )
             break;
-- 
2.1.0

